marshaling_rules.md

```diff
@@ -0,0 +1,121 @@
+# `R` <--> `Rust` marshaling rules
+
+This document outlines conversion and validation rules between `R` and `Rust` types.
+The aim is to maintain type safety on the `Rust` side without sacrificing usability on the `R` side.
+
+
+# Conversion and validation configuration
```

Comment:
Are these options necessary? Can't the configurations be determined uniquely by the type? I hope we can stick with `TryFrom` conversions for simplicity, if possible.

Reply:
I am sure they can, but we need to understand what types should be allowed in what regimes.

At some point when I was thinking about vectors, I realized there are three distinct problems that we must solve, which resulted in these configuration flags. 

When writing code, I prefer (and enjoy) two things: when syntactic sugar allows for simple, concise and readable code (with potential overhead under the hood), and when language features allow performant implementations using pointers and other tricks.
The first case is lazy `R` - `Rust` interop, when `Rust` function is called rarely, and provided input data are relatively small.
You implement `fn do_work(x : Vec<f64>)`, you call `do_work(1:100)`, and it just works.

However, if you want to speed up things and validate input yourself, you start using various `&[Rfloat]`, which just take ptr from `R` and use `std::slice::from_raw_parts`.